---
title: "Golden Gate TF reporter assembly - Sequence design"
author: "Max Trauernicht"
date: "`r format(Sys.time(), '%Y-%m-%d')`"
output: 
  html_document:
    theme: journal #cerulean
    highlight: monochrome
    toc: true
    toc_float: true
    code_folding: show
  editor_options:
    chunk_output_type: console
---

# knitr document van Steensel lab

# Introduction
Here, I will document how I designed the sequences for the tandem TF reporter array. 6 different TF reporters will be assembled, which contain BsmBI recognition sites in the 3' and 5' ends. By changing the 4 base overhangs created by the type IIs restriction enzyme, 5 different orders of the TF reporter assembly can be realized. 

## Description of Data
How to make a good rendering table: 
```{r table1, echo=FALSE, message=FALSE, warnings=FALSE, results='asis'}
tabl <- "
| column1	|	column2	|	column3	|	
|----|----|----|
|1	|	2	|	3	|	
|a	|	b	|	c	|	
"
cat(tabl) # output the table in a format good for HTML/PDF/docx conversion
```

# Data processing
## Path, Libraries, Parameters and Useful Functions
```{r setup}
knitr::opts_chunk$set(echo = TRUE)
StartTime <-Sys.time()

# 8-digit Date tag:
Date <- substr(gsub("-","",Sys.time()),1,8) 
# libraries:
library(ggplot2)
library(seqinr)
library(seqLogo)
library(universalmotif)
library(Biostrings)
library(SimRAD)
library(gtools)
library(DNABarcodes)
library(phylotools)
library(ape)
library(magrittr)
library(dplyr)
library(readr)
library(stringr)
library(tidyr)
library(heatmaply)
library(pheatmap)
library(tibble)
library(ggseqlogo)
library(RColorBrewer)
```

### Custom functions
Functions used thoughout this script.
```{r}
SetFileName <- function(filename, initials) {
  # Set filename with extension and initials to make filename with date integrated.
  filename <- substitute(filename)
  initials <- substitute(initials)
  filename <- paste0(initials, Date, filename)
  filename
}

# Function to substring the right part of the motif
substrRight <- function(x, n){
  substr(x, nchar(x)-n+1, nchar(x))
}

# Function to load PWM matrix
get_pwm_feature_matrix <- function(motif_meta_fn, fimo_fn, db = 2) {

  # validate args
  valid_dbs <- 1:2
  if(!db %in% valid_dbs)
    stop('Invalid db (database version). Please use db=1 (maintained for backward compatibility only) or db=2')

  # db=1 is maintained for backward compatibility only
  if(db == 1) {

    # read in motif metadata
    motif_meta    <- read.csv(motif_meta_fn)

    # check whether motif metadata contain essential annotations
    if(!all(c('PWM.ID', 'Cognate.TF') %in% colnames(motif_meta))) {
      message('The motif metadata file does not contain the essential columns PWM.ID and Cognate.TF')
    }

    motif_minimal <- motif_meta[, c('PWM.ID', 'Cognate.TF')]

    # load fimo output --> extract motif id, sequence id and p-value
    df <- read.table(fimo_fn)
    df <- df[, c(1, 2, 7)]

    colnames(df) <- c('PWM.ID', 'seqid', 'pval')

    # add TF id
    df <- merge(df, motif_minimal, by = 'PWM.ID')

    # group motif hits by sequence id
    l <- split(df, df[['seqid']])

    # multiple PWM and multiple hits possible. Reduce hits to one per TF, keeping best p-val only
    l <- lapply(l, function(x) {
      x_by_tf <- split(x, x[['Cognate.TF']], drop = TRUE)
      x_by_tf <- lapply(x_by_tf, function(y) y[which.min(y$pval), ])
      do.call('rbind', x_by_tf)
    })

    # initialize feature matrix
    n_tf          <- motif_minimal[['Cognate.TF']] %>%
      unique %>%
      length
    n_seq         <- length(l)
    pwm           <- matrix(1, nrow = n_seq, ncol = n_tf)
    colnames(pwm) <- (motif_minimal[['Cognate.TF']] %>% unique)

    # replace :: from names of composite motifs
    colnames(pwm) <- str_replace_all(colnames(pwm), '::', '_')

    # fill in feature matrix
    for(i in 1 : n_seq) {
      pwm[i, l[[i]][['Cognate.TF']]] <- l[[i]]$pval
    }

    # -log10 transform
    pwm           <- -1 * log10(pwm)

    # coerce to tib and return
    tib_fimo <- as_data_frame(pwm) %>%
      mutate(id = names(l))
      dplyr::select(id, everything())

  }

  # db = 2 (default)
  else {

    # load metadata
    tib_meta    <- read_csv(motif_meta_fn) %>%
      # extract tf symbol from motif id (Cognate_TF unsafe, it can be empty) and replace :: occurrences
      mutate(tf_symbol = str_remove(ID, '_[0-9]*'),
             tf_symbol = str_replace(tf_symbol, '::', '_')) %>%
      select(motif_id = `PWM ID`, tf_symbol)

    # load fimo results
    tib_fimo <- read_tsv(fimo_fn) %>%
      # extract motif id, sequence id and p-value
      select(motif_id, sequence_name, pval = `p-value`)

    # add tf symbol to fimo results
    tib_fimo <- tib_fimo %>%
      left_join(tib_meta, by = 'motif_id') %>%
      # remove hits with missing motif id (composite pwms)
      filter(!is.na(tf_symbol))

    # select best hit for each motif and sequence
    tib_fimo <- tib_fimo %>%
      group_by(sequence_name, tf_symbol) %>%
      dplyr::slice(which.min(pval)) %>%
      ungroup()

    # spread into feature matrix
    tib_fimo <- tib_fimo %>%
      mutate(pval = -1 * log10(pval)) %>%
      select(-motif_id) %>%
      spread(key = tf_symbol, value = pval, fill = 0, drop = TRUE) %>%
      # perform cosmetics on the id
      mutate(id = sequence_name) %>%
      select(-c(sequence_name)) %>%
      select(id, everything())

  }

  return(tib_fimo)

}
```


## Data import
```{r}
tf.array <- read.csv("/DATA/usr/m.trauernicht/projects/tf_activity_reporter/data/oligo_design/output/mt20191205_tf-array.csv")
```

# Analysis

## Selecting TF reporters
```{r}
# Select relevant TF reporters
tf.array$TF[tf.array$TF == "Nr3c1(GR)"]  <- "Gr"
tfs.selected <- c("^Stat3$", "^Trp53$", "^Random1$", "^Rara$", "^Gr$", "^Nfya$")
tf.array <- tf.array[grep(paste(tfs.selected, collapse = "|"),tf.array$TF),]
tf.array <- tf.array[tf.array$Promoter == "minP" & tf.array$Distance == "21bp" &
                       tf.array$Spacing == "10bp" & tf.array$Background == 1 &
                       tf.array$Barcode == 4,]

# Remove 5' and 3' primer adapter
tf.array <- tf.array %>% select(-Primer1_seq, -Primer2_seq)

# Add S2 primer adapter in front of pA signal - this is the sequence from the GFP gene
tf.array$S2_adapter <- "gccatgaagatcgagtgccgcatcacc"

# We can also change the S1 primer adapter to the full length
tf.array$S1_primer <- "acactctttccctacacgacgctcttccgatct"

# Attach pA site (SV-40)
tf.array$pA <- "aacttgtttattgcagcttataatggttacaaataaagcaatagcatcacaaatttcacaaataaagcatttttttcactgcattctagttgtggtttgtccaaactcatcaatgtatctta"
```


## Adding spacing in the transcriptional unit, between the barcode and the S2 adapter
```{r}
# Add 50 bp random inactive sequence in between barcode & pA site - otherwise the PCR product of the cDNA would be to small
## Generate 50bp oligos
# moderate GC content
# no TF binding in front of/after
x <- cbind.data.frame(c(1:200), c(""))
names(x) <- c("iteration", "oligo")


# Generate 2000 random 10bp oligos
# GC content ~50%
set.seed(8321233)
for (i in 1:200) {
  x$oligo[x$iteration == i] <- sim.DNAseq(50, GCfreq = 0.5)
}

# Generate all possible combinations of these motifs with primer sequences, spacers, etc.
tf.array2 <- merge(x, tf.array)

tf.array2$seq.name <- paste(tf.array2$TF, tf.array2$iteration, sep = "_")


# Assemble sequence to test: 4 rim bases of the TF motifs and the spacer sequence
tf.array2$seq.text <- paste(substrRight(tf.array2$barcode, 4), 
                       tf.array2$oligo, 
                       substr(tf.array2$S2_adapter, 1, 4), sep = "")

# Write fasta file to run on FIMO script
tf.array.export <- tf.array2 %>% select(seq.name, seq.text)
dat2fasta(tf.array.export, outfile = "/DATA/usr/m.trauernicht/projects/tf_activity_reporter/data/oligo_design/output/fimo/tf_array_GG_spacing.fasta")     
```

## Run FIMO script 
```{bash run fimo db2_9, eval = FALSE, echo=FALSE, message=FALSE, warnings=FALSE}
# motfn=/home/f.comoglio/mydata/Annotations/TFDB/Curated_Natoli/update_2017/20170320_pwms_selected.meme
# odir=/home/m.trauernicht/mydata/projects/tf_activity_reporter/data/oligo_design/output/fimo/tf_array_GG_spacing
# query=/home/m.trauernicht/mydata/projects/tf_activity_reporter/data/oligo_design/output/fimo/tf_array_GG_spacing.fasta

# nice -n 19 fimo --no-qvalue --thresh 1e-4 --verbosity 1 --o $odir $motfn $query 
```


## load fimo results
We built a TF motif matrix using -log10 transformed FIMO scores. We used this feature encoding throughout the rest of this analysis, unless otherwise stated.

```{r build tf motif matrices db2: primer_1, out.width= "100%", fig.align= "center", warning= FALSE}
# load motif Metadata --> PWM feature matrix
tib_pwm_tf_array_GG_spacer  <- get_pwm_feature_matrix(motif_meta_fn = '/home/f.comoglio/mydata/Annotations/TFDB/Curated_Natoli/update_2017/fc181127_curated_metadata_no_composite_filt.csv', 
                                        fimo_fn       = '/home/m.trauernicht/mydata/projects/tf_activity_reporter/data/oligo_design/output/fimo/tf_array_GG_spacing/fimo.tsv',
                                        db            = 2)
```




## visualize fimo results
```{r out.width= "100%", fig.align= "center", warning= FALSE, message = FALSE}
# convert to binary
tib_pwm_tf_array_GG_spacer_binary <- tib_pwm_tf_array_GG_spacer %>% mutate_if(is.numeric, function(x) ifelse((x==0 | is.na(x)),0,1)) 
tib_pwm_tf_array_GG_spacer_binary_top <- tib_pwm_tf_array_GG_spacer_binary

# compute rowsums to get cumulative binding
tib_pwm_tf_array_GG_spacer_binary_top$binding <- rowSums(tib_pwm_tf_array_GG_spacer_binary_top[,2:ncol(tib_pwm_tf_array_GG_spacer_binary_top)])

# select only cumulative binding and id
tib_pwm_tf_array_GG_spacer_binary_top <- tib_pwm_tf_array_GG_spacer_binary_top %>%
  select(id,binding)

# make third column wiht only the space_id - then sum up the binding scores per space_id
tib_pwm_tf_array_GG_spacer_binary_top$space <- gsub(".*_(.*)", "\\1", tib_pwm_tf_array_GG_spacer_binary_top$id)
tib_pwm_tf_array_GG_spacer_binary_top <- tib_pwm_tf_array_GG_spacer_binary_top %>%
  select(-id)
tib_pwm_tf_array_GG_spacer_binary_top$cum_binding <- ave(tib_pwm_tf_array_GG_spacer_binary_top$binding, tib_pwm_tf_array_GG_spacer_binary_top$space,
                                             FUN = sum)
tib_pwm_tf_array_GG_spacer_binary_top <- tib_pwm_tf_array_GG_spacer_binary_top %>%
  select(-binding)
tib_pwm_tf_array_GG_spacer_binary_top <- unique(tib_pwm_tf_array_GG_spacer_binary_top)

# Identify spacers that were exlcuded from fimo script (due to 0 hits)
space_id <- unique(tib_pwm_tf_array_GG_spacer_binary_top$space)
iteration <- 1:200
space_nohit <- iteration[! iteration %in% space_id]

# This revealed the best 2 hits - now add 7 more to complete 9 
tib_pwm_tf_array_GG_spacer_binary_top <- tib_pwm_tf_array_GG_spacer_binary_top[order(tib_pwm_tf_array_GG_spacer_binary_top$cum_binding),]

# Look at the top 3 to see which TF binds to them
low_binding_space <- tib_pwm_tf_array_GG_spacer_binary_top$space[1:3]

# Subselect top 12 from original df
low_binding_space2 <- merge(low_binding_space, unique(tf.array2$TF))


low_binding_space2$space <- paste(low_binding_space2$y, low_binding_space2$x, sep = "_")

tib_pwm_tf_array_GG_spacer_select <- tib_pwm_tf_array_GG_spacer[tib_pwm_tf_array_GG_spacer$id %in%
                                                                low_binding_space2$space,]



# convert to matrix - leave out id
tib_pwm_tf_array_GG_spacerMatrix <- as.matrix(select(tib_pwm_tf_array_GG_spacer_select,-id))

# assign ids as rownames of matrix
rownames(tib_pwm_tf_array_GG_spacerMatrix) <- tib_pwm_tf_array_GG_spacer_select$id
```

```{r, fig.height=10, fig.width=10, out.width= "100%", fig.align= "center", echo=FALSE, warning= FALSE}
#remove columns with only 0s and plot heatmap
heatmaply(tib_pwm_tf_array_GG_spacerMatrix[, colSums(tib_pwm_tf_array_GG_spacerMatrix != 0) > 0])
```

```{r}
# 50bp spacer 94 seems like a good candidate as it only allows Tcf7 binding
tf.array$transcription_spacer <- x$oligo[x$iteration == 94]
```








```{r}
# Concatenate intermediate sequence
tf.array$reporter <- paste(tf.array$motif1, tf.array$Space1, tf.array$motif2,
                           tf.array$Space2, tf.array$motif3, tf.array$Space3,
                           tf.array$motif4, tf.array$Distance_seq, tf.array$Promoter_sequence,
                           tf.array$S1_primer, tf.array$barcode, tf.array$transcription_spacer,
                           tf.array$S2_adapter, tf.array$pA, sep = "")
tf.array <- tf.array %>% select(TF, reporter)
```




# Add Golden Gate adapter sequences (BsmBI-based)
```{r}
# Select sequences and overlaps based on goldengate.neb.com using the tf.array sequences as input and the F3 exchange vector as backbone








# Add I-CeuI sites at the ends


# Check BsmBI restriction sites

# Check other critical restriction sites
```

# Conclusions
```{r}
```

## Exporting potential data. 
```{r}
```

# Session Info
```{r}
paste("Run time: ",format(Sys.time()-StartTime))
getwd()
date()
sessionInfo()
```

